orttraining/orttraining/training_api/onnxruntime_training_c_api.cc

```diff
@@ -0,0 +1,226 @@
+                    _In_ const ORTCHAR_T* train_model_path, _In_ const ORTCHAR_T* eval_model_path,
+                    _In_ const ORTCHAR_T* optimizer_model_path) {
+  API_IMPL_BEGIN
+
+  auto train_sess = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(session);
+  auto chkpt_state = reinterpret_cast<onnxruntime::training::api::CheckpointState*>(checkpoint_state);
+  ORT_API_RETURN_IF_STATUS_NOT_OK(train_sess->Initialize(chkpt_state->module_checkpoint_state.named_parameters,
+                                                         train_model_path,
+                                                         eval_model_path ? std::optional<std::string>{eval_model_path}
+                                                                         : std::nullopt,
+                                                         optimizer_model_path ? std::optional<std::string>{optimizer_model_path}
+                                                                              : std::nullopt));
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtApis::ResetGrad, _Inout_ OrtTrainingSession* session) {
+  API_IMPL_BEGIN
+  auto train_session = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(session);
+  ORT_API_RETURN_IF_STATUS_NOT_OK(train_session->ResetGrad());
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtApis::TrainStep, _Inout_ OrtTrainingSession* sess, _In_opt_ const OrtRunOptions* run_options,
+                  _In_reads_(inputs_len) const OrtValue* const* inputs, size_t inputs_len,
+                  size_t outputs_len, _Inout_updates_all_(outputs_len) OrtValue** outputs) {
+
+  API_IMPL_BEGIN
+  auto session = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(sess);
+  constexpr int queue_id = 0;
+
+  std::vector<OrtValue> feeds(inputs_len);
+
+  for (size_t i = 0; i != inputs_len; ++i) {
+    auto& ort_value = feeds[i] = *reinterpret_cast<const ::OrtValue*>(inputs[i]);
+    if (ort_value.Fence()) {
+      ort_value.Fence()->BeforeUsingAsInput(onnxruntime::kCpuExecutionProvider, queue_id);
+    }
+  }
+
+  // Create output feed
+  std::vector<OrtValue> fetches(outputs_len);
+  for (size_t i = 0; i != outputs_len; ++i) {
+    if (outputs[i] != nullptr) {
+      ::OrtValue& value = *(outputs[i]);
+      if (value.Fence())
+        value.Fence()->BeforeUsingAsOutput(onnxruntime::kCpuExecutionProvider, queue_id);
+      fetches[i] = value;
+    }
+  }
+  Status status;
+  if (run_options == nullptr) {
+    OrtRunOptions op;
+    status = session->TrainStep(op, feeds, fetches);
+  } else {
+    status = session->TrainStep(*run_options, feeds, fetches);
+  }
+
+  if (!status.IsOK())
+    return onnxruntime::ToOrtStatus(status);
+  for (size_t i = 0; i != outputs_len; ++i) {
+    ::OrtValue& value = fetches[i];
+    if (value.Fence())
+      value.Fence()->BeforeUsingAsInput(onnxruntime::kCpuExecutionProvider, queue_id);
+    if (outputs[i] == nullptr) {
```

Comment:
Need to think what the right behavior for training scenario should be... In case of inference users can allocate output buffers if the output shape is known and this code makes sense... If we dont expect the user to do this in training session then yes should return an error

Reply:
per offline discussion closing this .... 
Post Aishwarya's PR the only outputs will be the user outputs (and not gradient outputs) and in that case current code makes sense